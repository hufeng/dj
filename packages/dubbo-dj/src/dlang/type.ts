import { IType, TDataType } from '../types'
import { DubboDep } from './lang-deps'
import { DubboEntity } from './lang-entity'
import { DubboEnum } from './lang-enum'

export interface IProject {
  name: string
  version: string
  description: string
}

// ~~~~~~~~~ factory ~~~~~~~~~~~~~~~~~~~~~~~~~~
export function universalType(dep: DubboDep, type: TDataType) {
  if (type instanceof DubboEntity) {
    dep.add({
      fromModule: type.fullName,
      importModule: type.shortName,
    })
    return {
      tsType: type.infName,
      javaType: `((v:any) => v.__fields2java())`,
    }
  } else if (type instanceof DubboEnum) {
    dep.add({
      fromModule: type.fullName,
      importModule: type.shortName,
    })
    return {
      tsType: type.shortName,
      javaType: `((v:any) => java.enum('${type.fullName}', ${type.shortName}[v]))`,
    }
  } else {
    return type(dep)
  }
}

// ~~~~~~~~~ Boolean && boolean ~~~~~~~~~~~~~~~
export function Boolean(): IType {
  return {
    tsType: 'boolean',
    javaType: 'java.Boolean',
  }
}

export function boolean(): IType {
  return {
    tsType: 'boolean',
    javaType: 'java.boolean',
  }
}

// ~~~~~~~~~ each number type ~~~~~~~~~~~~~~
export function Integer(): IType {
  return {
    tsType: 'number',
    javaType: 'java.Integer',
  }
}

export function int(): IType {
  return {
    tsType: 'number',
    javaType: 'java.int',
  }
}

export function short(): IType {
  return {
    tsType: 'number',
    javaType: 'java.short',
  }
}

export function Short(): IType {
  return {
    tsType: 'number',
    javaType: 'java.Short',
  }
}

export function byte(): IType {
  return {
    tsType: 'number',
    javaType: 'java.byte',
  }
}

export function Byte(): IType {
  return {
    tsType: 'number',
    javaType: 'java.Byte',
  }
}

export function long(): IType {
  return {
    tsType: 'number',
    javaType: 'java.long',
  }
}

export function Long(): IType {
  return {
    tsType: 'number',
    javaType: 'java.Long',
  }
}

export function double(): IType {
  return {
    tsType: 'number',
    javaType: 'java.double',
  }
}

export function Double(): IType {
  return {
    tsType: 'number',
    javaType: 'java.Double',
  }
}

export function float(): IType {
  return {
    tsType: 'number',
    javaType: 'java.float',
  }
}

export function Float(): IType {
  return {
    tsType: 'number',
    javaType: 'java.Float',
  }
}

// ~~~~~~~~~~~~~~ String ~~~~~~~~~~~~~~~~
export function String(): IType {
  return {
    tsType: 'string',
    javaType: 'java.String',
  }
}

export function char(): IType {
  return {
    tsType: 'string',
    javaType: 'java.char',
  }
}

export function Char(): IType {
  return {
    tsType: 'string',
    javaType: 'java.Char',
  }
}

export function chars(): IType {
  return {
    tsType: 'string',
    javaType: 'java.chars',
  }
}

export function Character(): IType {
  return {
    tsType: 'string',
    javaType: 'java.Character',
  }
}

// ~~~~~~~~~~~~~~~~~~~~~ container ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
export function List(
  type: TDataType
): (() => IType) | ((dep: DubboDep) => IType) {
  if (type instanceof DubboEntity) {
    const { infName } = type
    return (dep: DubboDep) => {
      dep.add({
        fromModule: type.fullName,
        importModule: type.infName,
      })
      return {
        tsType: `Array<${infName}>`,
        javaType: `(function withList(val:Array<any>=[]) {return java.List(val.map(v => v.__fields2java()))})`,
      }
    }
  }

  if (type instanceof DubboEnum) {
    const { shortName, fullName } = type
    return (dep: DubboDep) => {
      dep.add({
        fromModule: type.fullName,
        importModule: type.shortName,
      })
      return {
        tsType: `Array<${shortName}>`,
        javaType: `(function withList(val:Array<any>=[]) {return java.List(val.map(v => java.enum('${fullName}')))})})`,
      }
    }
  }

  return (dep: DubboDep) => {
    const t = type(dep)
    return {
      tsType: `Array<${t.tsType}>`,
      javaType: `(function withList(val:Array<any>=[]){return java.List(val.map(v => ${t.javaType}(v)))})`,
    }
  }
}

/*
export function Set(type: Entity | LangEnum | (() => IType)): () => IType {
  return () => ({
    tsType: 'Array',
    javaType: 'java.Set',
    generic: [type],
  })
}

export function Collection(type: Entity | LangEnum | (() => IType)): () => IType {
  return () => ({
    tsType: 'Array',
    javaType: 'java.Collection',
    generic: [type],
  })
}

export function Iterator(type: Entity | LangEnum | (() => IType)): () => IType {
  return () => ({
    tsType: 'Array',
    javaType: 'java.Iterator',
    generic: [type],
  })
}

export function Enumeration(type: Entity | (() => IType)): () => IType {
  return () => ({
    tsType: 'Array',
    javaType: 'java.Iterator',
    generic: [type],
  })
}

export function HashMap(
  leftType: Entity | (() => IType),
  rightType: Entity | LangEnum | (() => IType)
): () => IType {
  return () => ({
    tsType: 'Map',
    javaType: 'java.HashMap',
    generic: [leftType, rightType],
  })
}

export function Map(
  leftType: Entity | (() => IType),
  rightType: Entity | LangEnum | (() => IType)
): () => IType {
  return () => ({
    tsType: 'Map',
    javaType: 'java.Map',
    generic: [leftType, rightType],
  })
}

export function Dictionary(
  leftType: Entity | (() => IType),
  rightType: Entity | (() => IType)
): () => IType {
  return () => ({
    tsType: 'Map',
    javaType: 'java.Dictionary',
    generic: [leftType, rightType],
  })
}
*/

export function Currency(): IType {
  return {
    tsType: 'string',
    javaType: 'java.Currency',
  }
}
