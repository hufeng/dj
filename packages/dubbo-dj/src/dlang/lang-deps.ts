export type TFromModule = string
export interface IDepValue {
  is3rdLib: boolean
  defaultModule: string
  noneDefaultModules: Set<string>
}
export interface IDepAdd {
  fromModule: string
  importModule: string
  isDefaultModule?: boolean
  is3rdLib?: boolean
}

export class DubboDep {
  private readonly depMap: Map<TFromModule, IDepValue>

  constructor() {
    this.depMap = new Map<TFromModule, IDepValue>()
  }

  add({
    fromModule,
    importModule,
    isDefaultModule = false,
    is3rdLib = false,
  }: IDepAdd) {
    if (this.depMap.has(fromModule)) {
      const moduleValue = this.depMap.get(fromModule) || {
        defaultModule: '',
        noneDefaultModules: new Set(),
      }
      if (isDefaultModule) {
        moduleValue.defaultModule = importModule
      } else {
        moduleValue.noneDefaultModules.add(importModule)
      }
    } else {
      if (isDefaultModule) {
        this.depMap.set(fromModule, {
          is3rdLib,
          defaultModule: importModule,
          noneDefaultModules: new Set<string>(),
        })
      } else {
        this.depMap.set(fromModule, {
          is3rdLib,
          defaultModule: '',
          noneDefaultModules: new Set([importModule]),
        })
      }
    }
  }

  getDepMap() {
    return this.depMap
  }
}
