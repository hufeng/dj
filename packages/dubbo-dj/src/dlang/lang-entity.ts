import Lang from './lang'
import { universalType } from './type'
import { IDubboEntityField, IType, TDataType } from '../types'

// ~~~~~~~~~~~~~~~~~~ model~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
export class DubboEntity extends Lang {
  fields: Array<IDubboEntityField>

  constructor(fullName: string, comment?: string) {
    super(fullName, comment)
    this.fields = []
    this.deps.add({
      fromModule: 'js-to-java',
      isDefaultModule: true,
      importModule: 'java',
      is3rdModule: true,
    })
  }

  field(name: string, type: IType, comment?: string) {
    this.fields.push({
      comment: this.wrapComment(comment),
      name,
      type,
    })
  }

  get toJSON() {
    return {
      mott: this.mott,
      comment: this.comment,
      fullName: this.fullName,
      shortName: this.shortName,
      infName: this.infName,
      fields: this.fields,
      imports: this.imports(this.deps),
    }
  }
}

// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ DSL ~~~~~~~~~~~~~~~~~~~~~~~
export class EntityBuilder {
  private readonly entity: DubboEntity

  constructor(fullName: string, comment?: string) {
    this.entity = new DubboEntity(fullName, comment)
  }

  field(name: string, type: TDataType, comment?: string) {
    const t = universalType(this.entity.deps, type)
    this.entity.field(name, t, comment)
    return this
  }

  ok() {
    return this.entity
  }
}

// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~ factory ~~~~~~~~~~~~~~~~~~~~~~~~
export function entity(fullName: string, comment?: string) {
  return new EntityBuilder(fullName, comment)
}
