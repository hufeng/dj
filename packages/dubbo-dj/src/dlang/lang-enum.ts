import Lang from './lang'
import { IDubboEnumFiled } from '../types'

// ~~~~~~~~~~~~~~~~~~ model~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
export class DubboEnum extends Lang {
  fields: Array<IDubboEnumFiled> = []

  constructor(fullName: string, comment?: string) {
    super(fullName, comment)
  }

  field(name: string, val: number | string, comment?: string) {
    this.fields.push({
      name,
      val,
      comment: this.wrapComment(comment),
    })
    return this
  }

  get toJSON() {
    return {
      mott: this.mott,
      comment: this.comment,
      fullName: this.fullName,
      fields: this.fields,
    }
  }
}

// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ DSL ~~~~~~~~~~~~~~~~~~~~~~~
export class DubboEnumDSL {
  private readonly enumeration: DubboEnum

  constructor(fullName: string, comment: string = '') {
    this.enumeration = new DubboEnum(fullName, comment)
  }

  /**
   * 枚举属性字段
   * @param name 属性名称
   * @param val
   * @param comment
   */
  field(name: string, val: number | string, comment?: string) {
    this.enumeration.field(name, val, comment)
    return this
  }

  /**
   * 完成Enum DSL的定义
   */
  ok() {
    return this.enumeration
  }
}

// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~ factory ~~~~~~~~~~~~~~~~~~~~~~~~
export function enumeration(cls: string, comment?: string) {
  return new DubboEnumDSL(cls, comment)
}
