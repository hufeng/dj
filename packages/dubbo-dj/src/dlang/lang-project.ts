import { IProject } from './type'

export class DubboProject {
  private readonly meta: IProject = { name: '', version: '', description: '' }

  name(name: string) {
    this.meta.name = name
    return this
  }

  version(version: string) {
    this.meta.version = version
    return this
  }

  description(description: string) {
    this.meta.description = description
    return this
  }

  ok() {
    return this.meta
  }
}

export const project = new DubboProject()
