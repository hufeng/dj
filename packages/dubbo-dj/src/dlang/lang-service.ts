import Lang from './lang'
import { universalType } from './type'
import { IArg, IFuncArg, IType, TDataType, TFuncOption } from '../types'
import { DubboDep } from './lang-deps'

// ~~~~~~~~~~~~~~~~~~~~~~~~~~~ model ~~~~~~~~~~~~~~~~~~~~~~~~~~
export class DubboService extends Lang {
  funcs: Array<Func>

  constructor(fullName: string, comment: string = '') {
    super(fullName, comment)
    this.funcs = []
  }

  get toJSON() {
    return {
      mott: this.mott,
      comment: this.wrapComment(this.comment),
      fullName: this.fullName,
      shortName: this.shortName,
      infName: this.infName,
      funcs: this.funcs,
      imports: this.imports(this.deps),
    }
  }
}

// ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ DSL ~~~~~~~~~~~~~~~~~~~~~~~
export class ServiceBuilder {
  private readonly service: DubboService

  constructor(fullName: string, comment: string = '') {
    this.service = new DubboService(fullName, comment)
  }

  func(...f: Array<TFuncOption>) {
    const func = new Func()
    for (let opt of f) {
      opt(func, this.service.deps)
    }
    this.service.funcs.push(func)
    return this
  }

  get ok() {
    return this.service
  }
}

export function service(name: string, comment?: string) {
  return new ServiceBuilder(name, comment)
}

// ~~~~~~~~~ Func && FuncBuilder ~~~~~~~~~~~~~~~~~~
export class Func {
  public comment: string = ''
  public name: string = ''
  public args: Array<IFuncArg> = []
  public ret: IType = null as any
}

// ~~~~~~~~~~~~~~~ functional option type ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
export const f = {
  name: (name: string, comment: string = '') => (func: Func) => {
    func.name = name
    func.comment = comment
  },
  ret: (type: TDataType) => (func: Func, deps: DubboDep) => {
    func.ret = universalType(deps, type)
  },
  args: (...args: Array<IArg>) => (func: Func, deps: DubboDep) => {
    func.args = args.map((v) => {
      return {
        ...v,
        type: universalType(deps, v.type),
      }
    })
  },
  arg: (name: string, type: TDataType, comment?: string) => {
    return {
      name,
      type,
      comment,
    }
  },
}
