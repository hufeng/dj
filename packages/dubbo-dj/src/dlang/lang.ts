import { DubboDep } from './lang-deps'

export default class Lang {
  fullName: string
  shortName: string
  // ts interface name
  infName: string

  deps: DubboDep
  comment: string

  constructor(fullClsName: string, comment?: string) {
    this.fullName = fullClsName
    this.shortName = fullClsName.split('.').pop() || ''
    this.infName = `I${this.shortName}`

    // init dep
    this.deps = new DubboDep()
    this.comment = this.wrapComment(comment)
  }

  wrapComment(comment: string = '') {
    return comment
      ? `
    /**
    * ${comment}
    */`
      : ''
  }

  // gen our mott
  get mott() {
    return `
      /**
       * auto generated by dubbo dj
       * ~~~ ðŸ’— machine coding ðŸ’— ~~~
       */
    `
  }

  imports(deps: DubboDep): string {
    const buff = []
    for (let [k, v] of deps.getDepMap().entries()) {
      buff.push('import ')
      if (v.defaultModule != '') {
        buff.push(` ${v.defaultModule} `)
        if (v.noneDefaultModules.size > 0) {
          buff.push(',')
        }
      }
      if (v.noneDefaultModules.size > 0) {
        buff.push(`{${[...v.noneDefaultModules].join(',')}}`)
      }
      buff.push(`from '${k}';`)
    }
    return buff.join('')
  }
}
