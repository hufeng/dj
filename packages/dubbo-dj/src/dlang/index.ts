import { entity } from './lang-entity'
import { service, f } from './lang-service'
import { enumeration } from './lang-enum'
import {
  Integer,
  String,
  Boolean,
  List,
  boolean,
  int,
  short,
  Short,
  byte,
  Byte,
  long,
  Long,
  double,
  Double,
  float,
  Float,
  char,
  Char,
  chars,
  Character,
  Currency,
} from './type'

export const t = {
  Boolean,
  boolean,
  Integer,
  int,
  short,
  Short,
  byte,
  Byte,
  long,
  Long,
  double,
  Double,
  float,
  Float,
  String,
  char,
  Char,
  chars,
  Character,
  List,
  Set,
  Currency,
}

export { entity, service, enumeration, f }
