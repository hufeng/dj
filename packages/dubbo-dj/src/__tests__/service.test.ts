import { entity, service, t } from '../dlang'
import { f } from '../dlang'
import {
  DubboAbstractServiceEmitter,
  DubboConsumerServiceEmitter,
  DubboServiceEmitter,
} from '../emitter'

describe('test s lang suite', () => {
  it('test basic s', () => {
    const addr = entity('org.apache.dubbo.domain.Address')
      .field('id', t.Integer, '地址id')
      .field('province', t.String, '省')
      .field('city', t.String, '市')
      .field('region', t.String, '区')
      .ok()
    const helloService = service('org.apache.dubbo.service.HelloService')
      .func(
        f.name('sayHello'),
        f.args(f.arg('name', t.String), f.arg('email', t.String)),
        f.ret(t.String)
      )
      .func(
        f.name('sayWorld'),
        f.args(f.arg('name', t.List(t.String)), f.arg('addr', t.List(addr))),
        f.ret(t.List(t.Integer))
      )
      .ok()

    expect(helloService).toMatchSnapshot()
    expect(
      new DubboServiceEmitter(helloService, 'ts').prettyCode
    ).toMatchSnapshot()

    expect(
      new DubboAbstractServiceEmitter(helloService, 'ts').prettyCode
    ).toMatchSnapshot()
    expect(
      new DubboConsumerServiceEmitter(helloService, 'ts').prettyCode
    ).toMatchSnapshot()
  })
})
