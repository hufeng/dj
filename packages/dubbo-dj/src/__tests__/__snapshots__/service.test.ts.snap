// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test s lang suite test basic s 1`] = `
DubboService {
  "comment": "",
  "deps": Dep {
    "depMap": Map {
      "org/apache/dubbo/domain/Address" => Object {
        "defaultModule": "",
        "is3rdModule": false,
        "noneDefaultModules": Set {
          "IAddress",
        },
      },
    },
  },
  "fullName": "org.apache.dubbo.service.HelloService",
  "funcs": Array [
    Func {
      "args": Array [
        Object {
          "comment": undefined,
          "name": "name",
          "type": Object {
            "javaType": "java.String",
            "tsInit": "",
            "tsType": "string",
          },
        },
        Object {
          "comment": undefined,
          "name": "email",
          "type": Object {
            "javaType": "java.String",
            "tsInit": "",
            "tsType": "string",
          },
        },
      ],
      "comment": "",
      "name": "sayHello",
      "ret": Object {
        "javaType": "java.String",
        "tsInit": "",
        "tsType": "string",
      },
    },
    Func {
      "args": Array [
        Object {
          "comment": undefined,
          "name": "name",
          "type": Object {
            "javaType": "((val:Array<string>=[]) => java.List(val.map(v => java.String(v))))",
            "tsInit": "",
            "tsType": "Array<string>",
          },
        },
        Object {
          "comment": undefined,
          "name": "addr",
          "type": Object {
            "javaType": "((list:Array<IAddress>=[]) => java.List(list.map(v => new Address(v).__fields2java())))",
            "tsInit": "((list:Array<IAddress>) => list.map(v => new Address(v)))",
            "tsType": "Array<IAddress>",
          },
        },
      ],
      "comment": "",
      "name": "sayWorld",
      "ret": Object {
        "javaType": "((val:Array<number>=[]) => java.List(val.map(v => java.Integer(v))))",
        "tsInit": "",
        "tsType": "Array<number>",
      },
    },
  ],
  "infName": "IHelloService",
  "modulePath": "org/apache/dubbo/service/HelloService",
  "shortName": "HelloService",
}
`;

exports[`test s lang suite test basic s 2`] = `
"/**
 * auto generated by dubbo dj
 * ~~~ ðŸ’— machine coding ðŸ’— ~~~
 */

import { IAddress } from '../domain/Address'
import HelloService from './base/HelloService'

export default class HelloServiceImpl extends HelloService {
  async sayHello(sayHello: string, sayHello: string): Promise<string> {
    throw new Error('Not implemented sayHello method')
  }

  async sayWorld(
    sayWorld: Array<string>,
    sayWorld: Array<IAddress>
  ): Promise<Array<number>> {
    throw new Error('Not implemented sayWorld method')
  }
}
"
`;

exports[`test s lang suite test basic s 3`] = `
"/**
 * auto generated by dubbo dj
 * ~~~ ðŸ’— machine coding ðŸ’— ~~~
 */

import java from 'js-to-java'
import { IAddress } from '../../domain/Address'

export default abstract class HelloService {
  dubboInterface = 'org.apache.dubbo.service.HelloService'

  methods = {
    sayHello: async (name: string, email: string) => {
      const val = await this.sayHello(name, email)
      return java.String(val)
    },
    sayWorld: async (name: Array<string>, addr: Array<IAddress>) => {
      const val = await this.sayWorld(name, addr)
      return ((val: Array<number> = []) =>
        java.List(val.map((v) => java.Integer(v))))(val)
    },
  }

  abstract sayHello(name: string, email: string): Promise<string>

  abstract sayWorld(
    name: Array<string>,
    addr: Array<IAddress>
  ): Promise<Array<number>>
}
"
`;

exports[`test s lang suite test basic s 4`] = `
"/**
 * auto generated by dubbo dj
 * ~~~ ðŸ’— machine coding ðŸ’— ~~~
 */

import { Dubbo, TDubboCallResult, java } from '@dubbo/dubbo-consumer'
import { IAddress } from '../domain/Address'

export interface IHelloService {
  sayHello(name: string, email: string): TDubboCallResult<string>
  sayWorld(
    name: Array<string>,
    addr: Array<IAddress>
  ): TDubboCallResult<Array<number>>
}

export function HelloService(dubbo: Dubbo): IHelloService {
  return dubbo.proxyService<IHelloService>({
    dubboInterface: 'org.apache.dubbo.service.HelloService',
    methods: {
      sayHello(name: string, email: string) {
        return [java.String(name), java.String(email)]
      },
      sayWorld(name: Array<string>, addr: Array<IAddress>) {
        return [
          ((val: Array<string> = []) =>
            java.List(val.map((v) => java.String(v))))(name),
          ((list: Array<IAddress> = []) =>
            java.List(list.map((v) => new Address(v).__fields2java())))(addr),
        ]
      },
    },
  })
}
"
`;
